// @instantdb/core/dist/commonjs/InMemoryStorage.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const PersistedObject_js_1 = require("./utils/PersistedObject.js");
class InMemoryStorage extends PersistedObject_js_1.StorageInterface {
    constructor(appId, dbName) {
        super(appId, dbName);
        this.store = new Map();
    }
    getItem(k) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            return (_a = this.store.get(k)) !== null && _a !== void 0 ? _a : null;
        });
    }
    setItem(k, v) {
        return __awaiter(this, void 0, void 0, function* () {
            this.store.set(k, v);
        });
    }
    getAllKeys() {
        return __awaiter(this, void 0, void 0, function* () {
            return [...this.store.keys()];
        });
    }
    multiSet(keyValuePairs) {
        return __awaiter(this, void 0, void 0, function* () {
            for (const [k, v] of keyValuePairs) {
                this.setItem(k, v);
            }
        });
    }
    removeItem(key) {
        return __awaiter(this, void 0, void 0, function* () {
            this.store.delete(key);
        });
    }
}
exports.default = InMemoryStorage;
